qutip/core/environment.py

```diff
@@ -0,0 +1,2015 @@
+        will be derived from the correlation function via a fast fourier
+        transform.
+
+        Parameters
+        ----------
+        w : array_like or float
+            The frequencies at which to evaluate the spectral density.
+        """
+
+        ...
+
+    @abc.abstractmethod
+    def correlation_function(
+        self, t: float | ArrayLike, *, eps: float = 1e-10
+    ) -> (float | ArrayLike):
+        """
+        The two-time auto-correlation function of this environment. See the
+        Users Guide on :ref:`bosonic environments <bosonic environments guide>`
+        for specifics on the definitions used by QuTiP.
+
+        If no analytical expression for the correlation function is known, it
+        will be derived from the power spectrum via a fast fourier transform.
+
+        If no analytical expression for the power spectrum is known either, it
+        will be derived from the spectral density. In this case, the
+        temperature of this environment must be specified.
+
+        Parameters
+        ----------
+        t : array_like or float
+            The times at which to evaluate the correlation function.
+
+        eps : optional, float
+            Used in case the power spectrum is derived from the spectral
+            density; see the documentation of
+            :meth:`BosonicEnvironment.power_spectrum`.
+        """
+
+        ...
+
+    @abc.abstractmethod
+    def power_spectrum(
+        self, w: float | ArrayLike, *, eps: float = 1e-10
+    ) -> (float | ArrayLike):
+        """
+        The power spectrum of this environment. See the Users Guide on
+        :ref:`bosonic environments <bosonic environments guide>` for specifics
+        on the definitions used by QuTiP.
+
+        If no analytical expression for the power spectrum is known, it will
+        be derived from the spectral density. In this case, the temperature of
+        this environment must be specified.
+
+        If no analytical expression for the spectral density is known either,
+        the power spectrum will instead be derived from the correlation
+        function via a fast fourier transform.
+
+        Parameters
+        ----------
+        w : array_like or float
+            The frequencies at which to evaluate the power spectrum.
+
+        eps : optional, float
+            To derive the zero-frequency power spectrum from the spectral
+            density, the spectral density must be differentiated numerically.
+            In that case, this parameter is used as the finite difference in
+            the numerical differentiation.
+        """
+
+        ...
+
+    # --- user-defined environment creation
+
+    @classmethod
+    def from_correlation_function(
+        cls,
+        C: Callable[[float], complex] | ArrayLike,
+        tlist: ArrayLike = None,
+        tMax: float = None,
+        *,
+        T: float = None,
+        tag: Any = None
```

Comment:
would be nice to be able to pass combine=True/False here too

Reply:
Did you mean to comment this somewhere else?

Reply:
Ah yeah, I meant for the cf_from_fit function, and anything else that returns an exponential environment.  Though I imagine we might adjust that at some point, but I like having control of the combine parameter in heom stuff